### ISSUE TITLE
Wrong handling for no_std?

### ISSUE THREAD

**Issue by Evian-Zhang** [m4b/goblin#234]
In `src/lib.rs`, I can see the following code:

```rust
#![cfg_attr(not(feature = "std"), no_std)]

#[cfg(feature = "std")]
extern crate core;
```

This seems to be very strange. In my understanding, `#![cfg_attr(not(feature = "std"), no_std)]` means using `#![no_std]` if `std` feature is not presented, and using `extern crate core` if `std` feature is presented.

However, as described in the [no_std](https://rust-embedded.github.io/book/intro/no-std.html) section of The Embedded Rust book:

> `#![no_std]` is a crate-level attribute that indicates that the crate will link to the core-crate instead of the std-crate.

And described in the [Using Rust Without the Standard Library](https://doc.rust-lang.org/1.30.0/book/first-edition/using-rust-without-the-standard-library.html) section in an outdated version of The Rust Programming Language book (current version has no description of the `#![no_std]` macro):

> When we’re using the standard library, Rust automatically brings std into scope, allowing you to use its features without an explicit import. By the same token, when using `#![no_std]`, Rust will bring `core` into scope for you, as well as its prelude.

So when we use `#![no_std]` macro, the `core` is automatically linked.

In conclusion, the code in `src/lib.rs` means importing `core` no matter `std` feature is presented or not, which really confuses me. Maybe we should use

```rust
#![cfg_attr(not(feature = "std"), no_std)]

#[cfg(feature = "std")]
use std;
```

instead?

**Comment by philipc** [comment]
We want to avoid doing things like this:

```rust
#[cfg(feature = "std")]
use std::mem;
#[cfg(not(feature = "std"))]
use core::mem;
```

There are a couple of ways to avoid that. One is to always import `core` and write `use core::mem;`. Another is to rename core for `no_std` builds with `use core as std;` and write `use std::mem;`, but I dislike that because it makes it harder to see which code really depends on `std`.

**Comment by Evian-Zhang** [comment]
OK, I get it.

Just to explain more explicitly for those who have the same question: the code in `src/lib.rs` is to make sure that:

* for those functionalities that both `std` and `core` has, we use `core` no matter `std` feature is presented or not;
* for those functionalities that only `std` has, we use `std` with `std` feature is presented;
* if `std` feature is not presented, we don't link `std` into this crate.

The code is the best way to implement the above requirements.
